### PR TITLE
chore(lsp): Add tracing to the deno LSP

### PR DESCRIPTION
**Pull request by nathanwhit**
Add tracing support to the LSP. Enabled via `deno.tracing.enable` in the editor settings or the `DENO_LSP_TRACE` environment variable. For instance, in your vscode `settings.json`, the following
```json
{
  "deno.tracing": {
    "enable": true,
    "filter": "debug"
   }
}
```
would enable tracing and filter to include all spans/events at a debug level or higher. By default, we use opentelemetry to export the data to a local opentelemetry collector. The easiest method is to just use [jaeger's all-in-one binary](https://www.jaegertracing.io/docs/1.56/getting-started/) to collect the tracing data, then open the jaeger UI to explore the data, for instance:
Screenshot 2024-04-04 at 2 28 31 PM

The biggest downside of this PR, as I see it, is the additional dependencies and presumably a corresponding increase in compile times. The additional code will probably also increase binary size, though I haven't measured how much. It may be worth feature gating this at some point so we don't always pay the cost.

--
To review this, you probably want to look at the first 3 commits, as that's where all of the actual changes are. The commit that adds `instrument` attributes to a bunch of functions is pretty noisy.

### CODE REVIEW

File: cli/lsp/trace.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+pub(crate) fn make_tracer(
+) -> Result<opentelemetry_sdk::trace::Tracer, anyhow::Error> {
+  Ok(
+    opentelemetry_otlp::new_pipeline()
+      .tracing()
+      .with_exporter(
+        opentelemetry_otlp::new_exporter()
+          .tonic()
+          .with_endpoint("http://localhost:4317"),
```
Comment (bartlomieju) [inline cli/lsp/trace.rs:27]: We probably shouldn't make it a fixed address?

**Review by bartlomieju — COMMENTED** [review]
Please add spans to all ops in `cli/lsp/tsc.rs` - I think we should collect module specifiers from these spans to further improve debugging.

**Comment by bartlomieju** [comment]
Tried id locally and I see a whole lot of messages like these in VSCode's "Output" tab for "Deno Language Server"
```
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
OpenTelemetry trace error occurred. cannot send message to batch processor as the channel is full
```

File: cli/lsp/trace.rs (hunk 1 above)
Comment (nathanwhit) [reply cli/lsp/trace.rs:27]: Yeah I'll add it to the config

File: cli/lsp/trace.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  #[default]
+  OpenTelemetry,
+  Logging,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, Default)]
+#[serde(default, rename_all = "camelCase")]
+pub(crate) struct TracingConfig {
+  /// Enable tracing.
+  pub(crate) enable: bool,
```
Comment (dsherret) [inline cli/lsp/trace.rs:78]: Maybe the precense of a tracing config should imply enablement? Then alternatively `"tracing": true`

**Review by ry — COMMENTED** [review]
The additional dependencies is unfortunate. Maybe use cargo feature flags and disable by default? Who do we expect to run this?

**Comment by bartlomieju** [comment]
> The additional dependencies is unfortunate. Maybe use cargo feature flags and disable by default? Who do we expect to run this?

We expect it for users to run by passing `deno.tracing` config in `vscode_deno`. If we feature flag it, it means users would have to compile Deno themselves to be able to leverage it. We should bite the bullet for now to improve the LSP and just accept these dependencies.

**Pull request closed without merging**